include/mqt-core/qdmi/sc/Device.hpp

```diff
@@ -0,0 +1,305 @@
+   */
+  explicit MQT_SC_QDMI_Device_Job_impl_d(
+      MQT_SC_QDMI_Device_Session_impl_d* session)
+      : session_(session) {}
+  /**
+   * @brief Frees the device job.
+   * @note This function just forwards to the session's @ref freeDeviceJob
+   * function. This function is needed because the interface only provides the
+   * job handle to the @ref QDMI_job_free function and the job's session handle
+   * is private.
+   * @see QDMI_job_free
+   */
+  auto free() -> void;
+
+  /**
+   * @brief Sets a parameter for the job.
+   * @see MQT_SC_QDMI_device_job_set_parameter
+   */
+  auto setParameter(QDMI_Device_Job_Parameter param, size_t size,
+                    const void* value) -> int;
+
+  /**
+   * @brief Queries a property of the job.
+   * @see MQT_SC_QDMI_device_job_query_property
+   */
+  auto queryProperty(QDMI_Device_Job_Property prop, size_t size, void* value,
+                     size_t* sizeRet) -> int;
+
+  /**
+   * @brief Submits the job to the device.
+   * @see MQT_SC_QDMI_device_job_submit
+   */
+  auto submit() -> int;
+
+  /**
+   * @brief Cancels the job.
+   * @see MQT_SC_QDMI_device_job_cancel
+   */
+  auto cancel() -> int;
+
+  /**
+   * @brief Checks the status of the job.
+   * @see MQT_SC_QDMI_device_job_check
+   */
+  auto check(QDMI_Job_Status* status) -> int;
+
+  /**
+   * @brief Waits for the job to complete but at most for the specified timeout.
+   * @see MQT_SC_QDMI_device_job_wait
+   */
+  auto wait(size_t timeout) -> int;
+
+  /**
+   * @brief Gets the results of the job.
+   * @see MQT_SC_QDMI_device_job_get_results
+   */
+  auto getResults(QDMI_Job_Result result, size_t size, void* data,
+                  [[maybe_unused]] size_t* sizeRet) -> int;
+};
+
+/**
+ * @brief Implementation of the MQT_SC_QDMI_Site structure.
+ */
+struct MQT_SC_QDMI_Site_impl_d {
+  friend MQT_SC_QDMI_Operation_impl_d;
+
+private:
+  uint64_t id_ = 0; ///< Unique identifier of the site
+
+  /**
+   * @brief Initializes a site implementation with the given unique identifier.
+   *
+   * @param id Unique identifier for the site.
+   */
+  explicit MQT_SC_QDMI_Site_impl_d(uint64_t id) : id_(id) {}
+
+public:
+  /// @brief Factory function for regular sites.
+  [[nodiscard]] static auto makeUniqueSite(uint64_t id)
+      -> std::unique_ptr<MQT_SC_QDMI_Site_impl_d>;
+  /**
+   * @brief Queries a property of the site.
+   * @see MQT_SC_QDMI_device_session_query_site_property
+   */
+  auto queryProperty(QDMI_Site_Property prop, size_t size, void* value,
+                     size_t* sizeRet) const -> int;
+};
+
+/**
+ * @brief Implementation of the MQT_SC_QDMI_Operation structure.
+ */
+struct MQT_SC_QDMI_Operation_impl_d {
+private:
+  std::string name_;     ///< Name of the operation
+  size_t numParameters_; ///< Number of parameters for the operation
+  /**
+   * @brief Number of qubits involved in the operation
+   */
+  std::optional<size_t> numQubits_ = std::nullopt;
```

Comment:
Why is this optional when the only constructor available requires the parameter to be set?